src/sqlfluff/core/parser/markers.py

```diff
@@ -1,190 +1,205 @@
-"""Implements the FilePositionMarker class.
+"""Implements the PositionMarker class.
 
 This class is a construct to keep track of positions within a file.
 """
 
-from typing import Optional
-
-
-class FilePositionMarker:
-    """This class is a construct to keep track of positions within a file."""
-
-    slots = ["statement_index", "line_no", "line_pos", "char_pos"]
-
-    def __init__(
-        self,
-        statement_index: Optional[int] = 1,
-        line_no: int = 1,
-        line_pos: Optional[int] = 1,
-        char_pos: int = 0,
-    ):
-        # NB: statement_index and line_pos are optional so that
-        # we can use position markers to flag up errors which
-        # are specific to a line but we don't (or can't) know
-        # precisely where within that line the issue is.
-        self.statement_index = statement_index
-        self.line_no = line_no
-        self.line_pos = line_pos
-        self.char_pos = char_pos
+from dataclasses import dataclass
+from typing import Tuple, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from sqlfluff.core.templaters import TemplatedFile
+
+
+@dataclass(frozen=True)
+class PositionMarker:
+    """A reference to a position in a file.
+
+    Things to note:
+        - This combines the previous functionality of FilePositionMarker
+          and EnrichedFilePositionMarker. Additionally it contains a reference
+          to the original templated file.
+        - It no longer explicitly stores a line number or line position in the
+          source or template. This is extrapolated from the templated file as required.
+        - Positions in the source and template are with slices and therefore identify
+          ranges.
+        - Positions within the fixed file are identified with a line number and line
+          position, which identify a point.
+          the fixed file, this is as a combination of line number and line position.
+        - Arithmetic comparisons are on the location in the fixed file.
+    """
+
+    source_slice: slice
```

Comment:
I assume `slice` is a type? I was briefly confused by a lower-case type, which is typically just built-in types such as `str` or `int`.

Reply:
`slice` is a built in type indeed. https://docs.python.org/3/glossary.html#term-slice